src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -721,43 +729,50 @@ impl Tree {
-        // 1. traverse the Tree, collect all useless tags in a Vec
-        // 2. traverse the Vec, remove all tags previously selected
-        // Bench it.
-        let children: SmallVec<_> = node
-            .children
-            .clone()
-            .into_iter()
-            .filter(|child| self.keep_only_needed(*child, live))
-            .collect();
-        let no_children = children.is_empty();
-        let node = self.nodes.get_mut(idx).unwrap();
-        node.children = children;
-        if !live.contains(&node.tag) && no_children {
-            // All of the children and this node are unreachable, delete this tag
-            // from the tree (the children have already been deleted by recursive
-            // calls).
-            // Due to the API of UniMap we must absolutely call
-            // `UniValMap::remove` for the key of this tag on *all* maps that used it
-            // (which are `self.nodes` and every range of `self.rperms`)
-            // before we can safely apply `UniValMap::forget` to truly remove
-            // the tag from the mapping.
-            let tag = node.tag;
-            self.nodes.remove(idx);
-            for (_perms_range, perms) in self.rperms.iter_mut_all() {
-                perms.remove(idx);
+    fn keep_only_needed(&mut self, root: UniIndex, live: &FxHashSet<BorTag>) -> bool {
+        let mut stack = vec![(root, 0)];
+        // To avoid stack overflows, we roll our own stack.
+        // Each frame consists of the current tag, and the number of the next child.
+        while let Some((tag, nth_child)) = stack.last_mut() {
+            let node = self.nodes.get(*tag).unwrap();
+            if *nth_child == node.children.len() {
+                // We have visited all children. Let's remove those that are useless.
+                // Split the children into useful and useless.
+                let (useless, new_children) =
+                    node.children.iter().copied().partition(|x| self.is_useless(*x, live));
+                // Assign the new children.
+                // This might make this node useless as well, but removal is the parent's job.
+                self.nodes.get_mut(*tag).unwrap().children = new_children;
+                // Next, ensure all useless children are properly unregistered everywhere.
+                for idx in useless {
+                    // Note: In the rest of this comment, "this node" means that at idx.
+                    // All of the children and this node are unreachable, delete this tag
+                    // from the tree (the children have already been deleted by recursive
+                    // calls).
+                    // Due to the API of UniMap we must absolutely call
+                    // `UniValMap::remove` for the key of this tag on *all* maps that used it
+                    // (which are `self.nodes` and every range of `self.rperms`)
+                    // before we can safely apply `UniValMap::forget` to truly remove
```

Comment:
Apparently this comment has gotten out of sync when we renamed `UniValMap::forget`
```suggestion
                    // before we can safely apply `UniValMap::remove` to truly remove
```